README.md
# RMQ - RabbitMQ tools.

[![Actions Status](https://github.com/nmbrone/rmq/workflows/CI/badge.svg)](https://github.com/nmbrone/rmq/actions)
[![Hex version badge](https://img.shields.io/hexpm/v/rmq.svg)](https://hex.pm/packages/rmq)

A set of handy tools for working with RabbitMQ in Elixir projects.
Based on [`AMQP`](https://github.com/pma/amqp) library.

It includes:

1. [RMQ.Connection](#rmqconnection)
2. [RMQ.Consumer](#rmqconsumer)
3. [RMQ.RPQ](#rmqrpc)

## Installation

The package can be installed by adding `rmq` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:rmq, "~> 0.3.0"}
  ]
end
```

## RMQ.Connection

A `GenServer` which provides a robust connection to the RabbitMQ server.

#### Usage

```elixir
defmodule MyApp.RabbitConnection do
  use RMQ.Connection,
    otp_app: :my_app,
    uri: "amqp://localhost",
    name: to_string(__MODULE__)
end
```

Meant to be started under the application's supervision tree as follows:

```elixir
defmodule MyApp.Application do
  use Application

  def start(_type, _args) do
    children = [
      MyApp.RabbitConnection
      # ...
    ]

    opts = [strategy: :one_for_one, name: MyApp.Supervisor]
    Supervisor.start_link(children, opts)
  end
end

```

#### Options

* `:uri` - an AMQP URI. Defaults to `"amqp://localhost"`;
* `:reconnect_interval` - a reconnect interval in milliseconds. It can be also a function that
  accepts the current connection attempt as a number and returns a new interval.
  Defaults to `5000`;
* other options for [`AMQP.Connection.open/2`](https://hexdocs.pm/amqp/1.5.0/AMQP.Connection.html#open/2).


## RMQ.Consumer

RabbitMQ Consumer.

#### Usage

```elixir
defmodule MyApp.Consumer do
  use RMQ.Consumer,
    queue: "my-app-consumer-queue",
    exchange: {"my-exchange", :direct, durable: true}

  @impl RMQ.Consumer
  def consume(chan, payload, meta) do
    # do something with the payload
    ack(chan, meta.delivery_tag)
  end
end

# or with dynamic configuration
defmodule MyApp.Consumer2 do
  use RMQ.Consumer

  @impl RMQ.Consumer
  def config do
    [
      queue: System.fetch_env!("QUEUE_NAME"),
      reconnect_interval: fn attempt -> attempt * 1000 end,
    ]
  end

  @impl RMQ.Consumer
  def consume(chan, payload, meta) do
    # do something with the payload
    ack(chan, meta.delivery_tag)
  end
end
```

#### Options

* `:connection` - the connection module which implements `RMQ.Connection` behaviour.
  Defaults to `RMQ.Connection`;
* `:queue` - the name of the queue to consume. Will be created if does not exist.
  Also can be a tuple `{queue, options}`. See the options for `AMQP.Queue.declare/3`;
* `:exchange` - the name of the exchange to which `queue` should be bound.
  Also can be a tuple `{type, exchange, options}`. See the options for
  `AMQP.Exchange.declare/4`. Defaults to `""`;
* `:routing_key` - queue binding key. Defaults to `queue`;
  Will be created if does not exist. Defaults to `""`;
* `:dead_letter` - defines if the consumer should setup deadletter exchange and queue.
  Defaults to `true`;
* `:dead_letter_queue` - the name of dead letter queue. Also can be a tuple `{queue, options}`.
  See the options for `AMQP.Queue.declare/3`. Defaults to `"#{queue}_error"`.;
* `:dead_letter_exchange` - the name of the exchange to which `dead_letter_queue` should be bound.
  Also can be a tuple `{type, exchange}` or `{type, exchange, options}`. See the options for
  `AMQP.Exchange.declare/4`. Defaults to `"#{exchange}.dead-letter"`;
* `:dead_letter_routing_key` - routing key for dead letter messages. Defaults to `queue`;
* `:prefetch_count` - sets the message prefetch count. Defaults to `10`;
* `:consumer_tag` - consumer tag. Defaults to a current module name;
* `:reconnect_interval` - a reconnect interval in milliseconds. It can be also a function that
  accepts the current connection attempt as a number and returns a new interval.
  Defaults to `5000`;

## RMQ.RPC

RPC via RabbitMQ.

#### Usage

```elixir
# Application 1:

defmodule MyApp.RemoteResource do
  use RMQ.RPC, publishing_options: [app_id: "MyApp"]

  def find_by_id(id) do
    call("remote-resource-finder", %{id: id})
  end
end

#  Application 2:

defmodule MyOtherApp.Consumer do
  use RMQ.Consumer, queue: "remote-resource-finder"

  @impl RMQ.Consumer
  def consume(chan, payload, meta) do
    response =
      payload
      |> Jason.decode!()
      |> Map.fetch!("id")
      |> MyOtherApp.Resource.get()
      |> Jason.encode!()

    reply(chan, meta, response)
    ack(chan, meta)
  end
end
```

#### Options

* `:connection` - the connection module which implements `RMQ.Connection` behaviour;
* `:queue` - the queue name to which the module will be subscribed for consuming responses.
  Also can be a tuple `{queue, options}`. See the options for `AMQP.Queue.declare/3`.
  Defaults to `""` which means the broker will assign a name to a newly created queue by itself;
* `:exchange` - the exchange name to which `:queue` will be bound.
  Please make sure the exchange exist. Defaults to `""` - the default exchange;
* `:consumer_tag` - a consumer tag for `:queue`. Defaults to the current module name;
* `:publishing_options` - any valid options for `AMQP.Basic.publish/5` except
  `reply_to`, `correlation_id` - these will be set automatically and cannot be changed.
  Defaults to `[]`;
* `:reconnect_interval` - a reconnect interval in milliseconds. It can be also a function that
  accepts the current connection attempt as a number and returns a new interval.
  Defaults to `5000`;
* `:filter_parameters` - a list of parameters that may contain sensitive data and have
  to be filtered out when logging. Defaults to `["password"]`.


CHANGELOG.md
# Changelog

## v0.5.0 (2020-09-28)

**Breaking changes:**

- Consumer: removed `concurrency` option.

**Improvements:**

- Logging.


test/rmq_test.exs
defmodule RMQTest do
  use ExUnit.Case
  doctest RMQ
end


test/support/case.ex
defmodule RMQ.Case do
  use ExUnit.CaseTemplate

  using do
    quote do
      use ExUnit.Case
      require RMQ.TestHelpers
      import RMQ.TestHelpers
    end
  end

  setup_all do
    start_supervised!(RMQ.TestConnection)
    :ok
  end

  setup do
    Process.register(self(), :current_test)
    :ok
  end
end


test/test_helper.exs
# Suppress RabbitMQ progress logs
:logger.add_primary_filter(
  :ignore_rabbitmq_progress_reports,
  {&:logger_filters.domain/2, {:stop, :equal, [:progress]}}
)

ExUnit.start()


mix.exs
defmodule RMQ.MixProject do
  use Mix.Project

  def project do
    [
      app: :rmq,
      version: "0.5.0",
      elixir: "~> 1.8",
      elixirc_paths: elixirc_paths(Mix.env()),
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      name: "RMQ",
      description: "RMQ - a set of tools for convenient work with RabbitMQ",
      source_url: "https://github.com/nmbrone/rmq",
      homepage_url: "https://github.com/nmbrone/rmq",
      package: [
        maintainers: ["Sergey Snozyk"],
        licenses: ["MIT"],
        links: %{"GitHub" => "https://github.com/nmbrone/rmq"}
      ],
      docs: [
        main: "RMQ",
        extras: ["README.md"]
      ]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:lager, :logger]
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:amqp, "~> 1.5"},
      {:ex_doc, "~> 0.21", only: :dev, runtime: false},
      {:jason, "~> 1.2"},
      {:uuid, "~> 1.1"}
    ]
  end
end


lib/rmq/rpc.ex
defmodule RMQ.RPC do
  @moduledoc """
  RPC via RabbitMQ.

  In short, it's a `GenServer` which implements a publisher and a consumer at once.

  You can read more about how this works in the
  [tutorial](https://www.rabbitmq.com/tutorials/tutorial-six-python.html).

  ## Configuration

    * `:connection` - the connection module which implements `RMQ.Connection` behaviour;
    * `:queue` - the queue name to which the module will be subscribed for consuming responses.
      Also can be a tuple `{queue, options}`. See the options for `AMQP.Queue.declare/3`.
      Defaults to `""` which means the broker will assign a name to the newly created queue by itself;
    * `:exchange` - the exchange name to which `:queue` will be bound.
      Please make sure the exchange exist. Defaults to `""` - the default exchange;
    * `:consumer_tag` - a consumer tag for `:queue`. Defaults to the current module name;
    * `:publishing_options` - any valid options for `AMQP.Basic.publish/5` except
      `reply_to`, `correlation_id` - these will be set automatically and cannot be changed.
      Defaults to `[]`;
    * `:reconnect_interval` - a reconnect interval in milliseconds. It can be also a function that
      accepts the current connection attempt as a number and returns a new interval.
      Defaults to `5000`;
    * `:filter_parameters` - a list of parameters that may contain sensitive data and have
      to be filtered out when logging. Defaults to `["password"]`.


  ## Example

  Application 1:

      defmodule MyApp.RemoteResource do
        use RMQ.RPC, publishing_options: [app_id: "MyApp"]

        def find_by_id(id) do
          call("remote-resource-finder", %{id: id})
        end
      end

  Application 2:

      defmodule MyOtherApp.Consumer do
        use RMQ.Consumer, queue: "remote-resource-finder"

        @impl RMQ.Consumer
        def consume(chan, payload, meta) do
          response =
            payload
            |> Jason.decode!()
            |> Map.fetch!("id")
            |> MyOtherApp.Resource.get()
            |> Jason.encode!()

          reply(chan, meta, response)
          ack(chan, meta)
        end
      end

  """

  use RMQ.Logger
  import RMQ.Utils

  @defaults [
    connection: RMQ.Connection,
    queue: "",
    exchange: "",
    publishing_options: [],
    reconnect_interval: 5000,
    filter_parameters: ["password"]
  ]

  @doc """
  A callback for dynamic configuration.

  Will be called before `c:setup_queue/2`.
  """
  @callback config() :: keyword()

  @doc """
  Does all the job on preparing the queue.

  Whenever you need full control over configuring the queue you can implement this callback and
  use `AMQP` library directly.

  See `setup_queue/2` for the default implementation.
  """
  @callback setup_queue(chan :: AMQP.Channel.t(), config :: keyword()) :: binary()

  @doc false
  def start_link(module, opts) do
    GenServer.start_link(module, module, Keyword.put_new(opts, :name, module))
  end

  @doc """
  Performs a call against the given module.

  Here `options` is a keyword list which will be merged into `:publishing_options` during the call
  and `timeout` is the timeout in milliseconds for the inner GenServer call.

  The payload will be encoded by using `Jason.encode!/1`.
  """
  @spec call(
          module :: module(),
          queue :: binary(),
          payload :: any(),
          options :: keyword(),
          timeout :: integer()
        ) :: {:ok, any()} | {:error, :not_connected} | {:error, :timeout} | {:error, any()}
  def call(module, queue, payload \\ %{}, options \\ [], timeout \\ 5000) do
    GenServer.call(module, {:publish, queue, payload, options}, timeout)
  catch
    :exit, {:timeout, _} -> {:error, :timeout}
  end

  @doc """
  The default implementation for `c:setup_queue/2` callback.
  """
  @spec setup_queue(chan :: AMQP.Channel.t(), config :: keyword()) :: binary()
  def setup_queue(chan, conf) do
    {q, opts} = normalize_queue(conf[:queue])

    {:ok, %{queue: queue}} =
      AMQP.Queue.declare(chan, q, Keyword.merge([exclusive: true, auto_delete: true], opts))

    unless conf[:exchange] == "" do
      :ok = AMQP.Queue.bind(chan, queue, conf[:exchange], routing_key: queue)
    end

    {:ok, _} = AMQP.Basic.consume(chan, queue, nil, consumer_tag: conf[:consumer_tag])
    queue
  end

  @doc false
  def init(_module, _arg) do
    Process.flag(:trap_exit, true)
    send(self(), {:init, 1})
    {:ok, %{chan: nil, queue: nil, pids: %{}, config: %{}}}
  end

  @doc false
  def handle_call(_module, {:publish, _queue, _payload, _options}, _from, %{chan: nil} = state) do
    {:reply, {:error, :not_connected}, state}
  end

  def handle_call(module, {:publish, queue, payload, options}, from, %{config: config} = state) do
    correlation_id = UUID.uuid1()

    options =
      config[:publishing_options]
      |> Keyword.merge(options)
      |> Keyword.put(:reply_to, state.queue)
      |> Keyword.put(:correlation_id, correlation_id)

    log_debug(module, [
      "Publishing >>> ",
      "queue=#{inspect(queue)} ",
      "payload=#{inspect(filter_values(payload, config[:filter_parameters]))}"
    ])

    payload = encode_message(payload)

    case AMQP.Basic.publish(state.chan, config[:exchange], queue, payload, options) do
      :ok -> {:noreply, put_in(state.pids[correlation_id], from)}
      {:error, reason} -> {:reply, {:error, reason}, state}
    end
  end

  @doc false
  def handle_info(module, {:init, attempt}, state) do
    config = module_config(module)

    with {:ok, conn} <- config[:connection].get_connection(),
         {:ok, chan} <- AMQP.Channel.open(conn) do
      Process.monitor(chan.pid)
      queue = module.setup_queue(chan, config)
      log_info("Ready")
      {:noreply, %{state | chan: chan, queue: queue, config: config}}
    else
      error ->
        time = reconnect_interval(config[:reconnect_interval], attempt)
        log_error("No connection: #{inspect(error)}. Retrying in #{time}ms")
        Process.send_after(self(), {:init, attempt + 1}, time)
        {:noreply, %{state | config: config}}
    end
  end

  # Confirmation sent by the broker after registering this process as a consumer
  def handle_info(_module, {:basic_consume_ok, _meta}, state) do
    {:noreply, state}
  end

  # Sent by the broker when the consumer is unexpectedly cancelled (such as after a queue deletion)
  def handle_info(_module, {:basic_cancel, _meta}, state) do
    {:stop, :normal, state}
  end

  # Confirmation sent by the broker to the consumer process after a Basic.cancel
  def handle_info(_module, {:basic_cancel_ok, _meta}, state) do
    {:noreply, state}
  end

  def handle_info(module, {:basic_deliver, payload, meta}, %{config: config} = state) do
    {pid, state} = pop_in(state.pids[meta.correlation_id])

    unless is_nil(pid) do
      payload = decode_message(payload)

      log_debug(module, [
        "Consuming <<< ",
        "payload=#{inspect(filter_values(payload, config[:filter_parameters]))}"
      ])

      GenServer.reply(pid, {:ok, payload})
      AMQP.Basic.ack(state.chan, meta.delivery_tag)
    end

    {:noreply, state}
  end

  def handle_info(module, {:DOWN, _ref, :process, _pid, reason}, state) do
    log_error(module, "Connection lost: #{inspect(reason)}. Reconnecting...")
    send(self(), {:init, 1})
    {:noreply, %{state | chan: nil}}
  end

  def handle_info(module, :shutdown = reason, state) do
    terminate(module, reason, state)
    {:noreply, state}
  end

  @doc false
  def terminate(_module, _reason, %{chan: chan}) do
    close_channel(chan)
  end

  defp module_config(module) do
    @defaults
    |> Keyword.merge(module.config())
    |> Keyword.put_new(:consumer_tag, to_string(module))
  end

  defmacro __using__(opts \\ []) do
    quote location: :keep do
      use GenServer

      @behaviour RMQ.RPC
      @config unquote(opts)

      def start_link(opts), do: RMQ.RPC.start_link(__MODULE__, opts)

      def call(queue, payload \\ %{}, options \\ [], timeout \\ 5000) do
        RMQ.RPC.call(__MODULE__, queue, payload, options, timeout)
      end

      @impl RMQ.RPC
      def config, do: @config

      @impl RMQ.RPC
      def setup_queue(chan, config), do: RMQ.RPC.setup_queue(chan, config)

      @impl GenServer
      def init(arg), do: RMQ.RPC.init(__MODULE__, arg)

      @impl GenServer
      def handle_call(msg, from, state), do: RMQ.RPC.handle_call(__MODULE__, msg, from, state)

      @impl GenServer
      def handle_info(msg, state), do: RMQ.RPC.handle_info(__MODULE__, msg, state)

      @impl GenServer
      def terminate(reason, state), do: RMQ.RPC.terminate(__MODULE__, reason, state)

      defoverridable config: 0, setup_queue: 2
    end
  end
end


lib/rmq/utils.ex
defmodule RMQ.Utils do
  @moduledoc """
  Utility functions.
  """

  @doc false
  def reconnect_interval(int, _attempt) when is_integer(int), do: int
  def reconnect_interval(func, attempt) when is_function(func), do: func.(attempt)

  @doc false
  def close_channel(%AMQP.Channel{pid: pid} = chan) do
    if Process.alive?(pid), do: AMQP.Channel.close(chan), else: :ok
  end

  def close_channel(_), do: :ok

  @doc false
  def normalize_queue({queue, opts}), do: {queue, opts}
  def normalize_queue(queue), do: {queue, []}

  @doc false
  def normalize_exchange({name, type, opts}), do: {name, type, opts}
  def normalize_exchange({type, name}), do: {name, type, []}
  def normalize_exchange(name), do: {name, :direct, []}

  @doc false
  def encode_message(message) when is_binary(message), do: message

  def encode_message(message) do
    case Jason.encode(message) do
      {:ok, encoded} -> encoded
      {:error, _} -> message
    end
  end

  @doc false
  def decode_message(message) do
    case Jason.decode(message) do
      {:ok, decoded} -> decoded
      {:error, _} -> message
    end
  end

  @doc """
  Acknowledges one or more messages.

  It's the same as `AMQP.Basic.ack/3` but with the ability to extract `delivery_tag` from
  the provided message meta.
  """
  @spec ack(
          chan :: AMQP.Channel.t(),
          delivery_tag_or_meta :: binary() | map(),
          options :: keyword()
        ) :: :ok | (error :: any())
  def ack(chan, delivery_tag_or_meta, options \\ [])

  def ack(chan, %{delivery_tag: delivery_tag}, options) do
    AMQP.Basic.ack(chan, delivery_tag, options)
  end

  def ack(chan, delivery_tag, options) do
    AMQP.Basic.ack(chan, delivery_tag, options)
  end

  @doc """
  Produces a reply to the message.

  It's basically just a wrapper around `AMQP.Basic.publish/5` which helps you properly publish
  a reply based on the message meta.

  When the given payload is not a string it will be encoded by using `Jason.encode/2`.
  """
  @spec reply(chan :: AMQP.Channel.t(), meta :: map(), payload :: any(), options :: keyword()) ::
          :ok | (error :: any())
  def reply(chan, meta, payload, options \\ [])
  def reply(_, %{reply_to: :undefined}, _, _), do: :missing_repty_to
  def reply(_, %{correlation_id: :undefined}, _, _), do: :missing_correlation_id

  def reply(chan, %{exchange: xch, reply_to: to, correlation_id: cid}, payload, options) do
    payload = encode_message(payload)
    options = Keyword.put(options, :correlation_id, cid)
    AMQP.Basic.publish(chan, xch, to, payload, options)
  end

  def reply(_, _, _, _), do: :error

  @doc false
  def filter_values(%{__struct__: mod} = struct, _params) when is_atom(mod) do
    struct
  end

  def filter_values(map, params) when is_map(map) do
    Enum.into(map, %{}, &do_filter(&1, params))
  end

  def filter_values(list, params) when is_list(list) do
    if Keyword.keyword?(list) do
      Enum.into(list, [], &do_filter(&1, params))
    else
      Enum.map(list, &filter_values(&1, params))
    end
  end

  def filter_values(other, _params), do: other

  defp do_filter({k, v}, params) do
    if k in params do
      {k, "[FILTERED]"}
    else
      {k, filter_values(v, params)}
    end
  end
end


lib/rmq/logger.ex
defmodule RMQ.Logger do
  defmacro __using__(_) do
    quote do
      require Logger

      @log_prefix "[#{inspect(__MODULE__)}]"

      defp log_debug(message), do: Logger.debug("#{@log_prefix} #{message}")
      defp log_info(message), do: Logger.info("#{@log_prefix} #{message}")
      defp log_warn(message), do: Logger.warn("#{@log_prefix} #{message}")
      defp log_error(message), do: Logger.error("#{@log_prefix} #{message}")

      defp log_debug(module, message), do: Logger.debug("[#{inspect(module)}] #{message}")
      defp log_info(module, message), do: Logger.info("[#{inspect(module)}] #{message}")
      defp log_warn(module, message), do: Logger.warn("[#{inspect(module)}] #{message}")
      defp log_error(module, message), do: Logger.error("[#{inspect(module)}] #{message}")
    end
  end
end


lib/rmq/connection.ex
defmodule RMQ.Connection do
  @moduledoc ~S"""
  A `GenServer` which provides a robust connection to the RabbitMQ server.

  ## Usage

      iex> RMQ.Connection.start_link([])
      {:ok, #PID<0.310.0>}
      iex> RMQ.Connection.get_connection()
      {:ok, %AMQP.Connection{pid: #PID<0.314.0>}}

  ## Configuration

      config :rmq, :connection,
        uri: "amqp://localhost",
        name: "MyAppConnection",
        reconnect_interval: 5000,
        username: "user",
        password: "password"
        # ... other options for AMQP.Connection.open/3

    All configuration is optional.

    * `:uri` - an AMQP URI. Defaults to `"amqp://localhost"`;
    * `:reconnect_interval` - a reconnect interval in milliseconds. It can be also a function that
      accepts the current connection attempt as a number and returns a new interval.
      Defaults to `5000`;
    * other options for `AMQP.Connection.open/2`.

  ## Dynamic configuration

  In case you need to read the configuration dynamically you can use `c:config/0` callback:

      defmodule MyApp.RabbitConnection do
        use RMQ.Connection

        def config do
          [
            uri: System.get_env("RABBITMQ_URI"),
            name: "MyAppConnection",
            reconnect_interval: fn attempt -> attempt * 1000 end
            # ...
          ]
        end
      end

  ## Multiple connections

  If for some reason, you need to hold multiple connections you can use the following approach:

      defmodule MyApp.RabbitConnection1 do
        use RMQ.Connection, otp_app: :my_app
      end

      defmodule MyApp.RabbitConnection2 do
        use RMQ.Connection, otp_app: :my_app
      end

      # config.exs
      config :my_app, MyApp.RabbitConnection1,
        uri: "amqp://localhost",
        name: "MyAppConnection1"

      config :my_app, MyApp.RabbitConnection2,
        uri: "amqp://localhost",
        name: "MyAppConnection2"

  `otp_app: :my_app` here can be omitted and in that case `otp_app: :rmq` will be used.
  """

  use GenServer
  use RMQ.Logger

  @doc """
  A callback invoked right before connection.

  `use RMQ.Connection` will inject the default implementation of it:

      def config do
        Application.get_env(@otp_app, __MODULE__, [])
      end

  What can be eventually overridden into something like:

      def config do
        Keyword.merge(super(), uri: System.get_env("RABBITMQ_URI"))
      end
  """
  @callback config() :: keyword()

  @doc "Starts a `GenServer` process linked to the current process."
  def start_link(opts), do: start_link(__MODULE__, opts)

  @doc "Starts a `GenServer` process linked to the current process."
  def start_link(module, opts) do
    GenServer.start_link(__MODULE__, module, Keyword.put(opts, :name, module))
  end

  @doc "Gets the connection."
  @spec get_connection(module :: module()) ::
          {:ok, AMQP.Connection.t()} | {:error, :not_connected}
  def get_connection(module \\ __MODULE__) do
    case GenServer.call(module, :get) do
      nil -> {:error, :not_connected}
      conn -> {:ok, conn}
    end
  end

  @doc "Returns the configuration."
  def config(), do: Application.get_env(:rmq, :connection, [])

  @impl GenServer
  def init(module) do
    Process.flag(:trap_exit, true)
    send(self(), {:connect, 1})
    {:ok, %{module: module, conn: nil}}
  end

  @impl GenServer
  def handle_call(:get, _from, state) do
    {:reply, state.conn, state}
  end

  @impl GenServer
  def handle_info({:connect, attempt}, %{module: module} = state) do
    {uri, options} = Keyword.pop(module.config(), :uri, "amqp://localhost")
    {reconnect_interval, options} = Keyword.pop(options, :reconnect_interval, 5000)

    case AMQP.Connection.open(uri, options) do
      {:ok, conn} ->
        log_info("Connected")
        Process.monitor(conn.pid)
        {:noreply, %{state | conn: conn}}

      {:error, reason} ->
        ms = RMQ.Utils.reconnect_interval(reconnect_interval, attempt)
        log_error("Connection failed: #{inspect(reason)}. Reconnecting in #{ms}ms")
        Process.send_after(self(), {:connect, attempt + 1}, ms)
        {:noreply, state}
    end
  end

  def handle_info({:DOWN, _ref, :process, _pid, reason}, %{module: module} = state) do
    log_error(module, "Connection lost: #{inspect(reason)}. Reconnecting...")
    send(self(), {:connect, 1})
    {:noreply, %{state | conn: nil}}
  end

  @impl GenServer
  def terminate(_reason, %{conn: conn}) do
    unless is_nil(conn), do: AMQP.Connection.close(conn)
    :ok
  end

  @doc false
  defmacro __using__(opts \\ []) do
    quote location: :keep, bind_quoted: [opts: opts] do
      @behaviour RMQ.Connection
      @otp_app Keyword.get(opts, :otp_app, :rmq)

      def child_spec(opts), do: %{id: __MODULE__, start: {__MODULE__, :start_link, [opts]}}

      def start_link(opts), do: RMQ.Connection.start_link(__MODULE__, opts)

      def get_connection(), do: RMQ.Connection.get_connection(__MODULE__)

      @impl RMQ.Connection
      def config(), do: Application.get_env(@otp_app, __MODULE__, [])

      defoverridable config: 0
    end
  end
end


lib/rmq/consumer.ex
defmodule RMQ.Consumer do
  @moduledoc ~S"""
  RabbitMQ Consumer.

  ## Configuration

    * `:connection` - the connection module which implements `RMQ.Connection` behaviour.
      Defaults to `RMQ.Connection`;
    * `:queue` - the name of the queue to consume. Will be created if does not exist.
      Also can be a tuple `{queue, options}`. See the options for `AMQP.Queue.declare/3`;
    * `:exchange` - the name of the exchange to which `queue` should be bound.
      Also can be a tuple `{exchange, type, options}`. See the options for
      `AMQP.Exchange.declare/4`. Defaults to `""`;
    * `:routing_key` - queue binding key. Defaults to `queue`;
      Will be created if does not exist. Defaults to `""`;
    * `:dead_letter` - defines if the consumer should setup deadletter exchange and queue.
      Defaults to `true`;
    * `:dead_letter_queue` - the name of dead letter queue. Also can be a tuple `{queue, options}`.
      See the options for `AMQP.Queue.declare/3`. Defaults to `"#{queue}_error"`.;
    * `:dead_letter_exchange` - the name of the exchange to which `dead_letter_queue` should be bound.
      Also can be a tuple `{type, exchange}` or `{type, exchange, options}`. See the options for
      `AMQP.Exchange.declare/4`. Defaults to `"#{exchange}.dead-letter"`;
    * `:dead_letter_routing_key` - routing key for dead letter messages. Defaults to `queue`;
    * `:prefetch_count` - sets the message prefetch count. Defaults to `10`;
    * `:consumer_tag` - consumer tag. Defaults to a current module name;
    * `:reconnect_interval` - a reconnect interval in milliseconds. It can be also a function that
      accepts the current connection attempt as a number and returns a new interval.
      Defaults to `5000`;

  ## Examples

      defmodule MyApp.Consumer do
        use RMQ.Consumer,
          queue: {"my-app-consumer-queue", durable: true},
          exchange: {"my-exchange", :direct, durable: true}

        @impl RMQ.Consumer
        def consume(chan, payload, meta) do
          # do something with the payload
          ack(chan, meta.delivery_tag)
        end
      end

      defmodule MyApp.Consumer2 do
        use RMQ.Consumer

        @impl RMQ.Consumer
        def config do
          [
            queue: System.fetch_env!("QUEUE_NAME"),
            reconnect_interval: fn attempt -> attempt * 1000 end,
          ]
        end

        @impl RMQ.Consumer
        def consume(chan, payload, meta) do
          # do something with the payload
          ack(chan, meta.delivery_tag)
        end
      end

  """

  use RMQ.Logger
  import RMQ.Utils

  @defaults [
    connection: RMQ.Connection,
    exchange: "",
    dead_letter: true,
    prefetch_count: 10,
    reconnect_interval: 5000
  ]

  @doc """
  A callback for consuming a message.

  Keep in mind that the consumed message needs to be explicitly acknowledged via `AMQP.Basic.ack/3`
  or rejected via `AMQP.Basic.reject/3`. For convenience, these functions
  are imported and are available for use directly.

  `RMQ.Utils.reply/4` is imported as well which is convenient for the case when the consumer
  implements RPC.
  """
  @callback consume(chan :: AMQP.Channel.t(), payload :: any(), meta :: map()) :: any()

  @doc """
  A callback for dynamic configuration.
  """
  @callback config() :: keyword()

  @doc """
  Does all the job on preparing the queue.

  Whenever you need full control over configuring the queue you can implement this callback and
  use `AMQP` library directly.

  See `setup_queue/2` for the default implementation.
  """
  @callback setup_queue(chan :: AMQP.Channel.t(), config :: keyword()) :: :ok

  @doc false
  def start_link(module, opts) do
    GenServer.start_link(module, module, Keyword.put_new(opts, :name, module))
  end

  @doc false
  def init(_module, _arg) do
    Process.flag(:trap_exit, true)
    send(self(), {:init, 1})
    {:ok, %{chan: nil, config: %{}}}
  end

  @doc false
  def handle_info(module, {:init, attempt}, state) do
    config = module_config(module)

    with {:ok, conn} <- config[:connection].get_connection(),
         {:ok, chan} <- AMQP.Channel.open(conn) do
      Process.monitor(chan.pid)
      module.setup_queue(chan, config)
      log_info("Ready")
      {:noreply, %{state | config: config, chan: chan}}
    else
      error ->
        time = reconnect_interval(config[:reconnect_interval], attempt)
        log_error("No connection: #{inspect(error)}. Retrying in #{time}ms")
        Process.send_after(self(), {:init, attempt + 1}, time)
        {:noreply, %{state | config: config}}
    end
  end

  # Confirmation sent by the broker after registering this process as a consumer
  def handle_info(_module, {:basic_consume_ok, _meta}, state) do
    {:noreply, state}
  end

  # Sent by the broker when the consumer is unexpectedly cancelled (such as after a queue deletion)
  def handle_info(_module, {:basic_cancel, _meta}, state) do
    {:stop, :normal, state}
  end

  # Confirmation sent by the broker to the consumer process after a Basic.cancel
  def handle_info(_module, {:basic_cancel_ok, _meta}, state) do
    {:noreply, state}
  end

  def handle_info(module, {:basic_deliver, payload, meta}, %{chan: chan} = state) do
    module.consume(chan, payload, meta)
    {:noreply, state}
  end

  def handle_info(module, {:DOWN, _ref, :process, _pid, reason}, state) do
    log_error(module, "Connection lost: #{inspect(reason)}. Reconnecting...")
    send(self(), {:init, 1})
    {:noreply, %{state | chan: nil}}
  end

  @doc false
  def terminate(_module, _reason, %{chan: chan}) do
    close_channel(chan)
  end

  @doc """
  The default implementation for `c:setup_queue/2` callback.
  """
  @spec setup_queue(chan :: AMQP.Channel.t(), config :: keyword()) :: :ok
  def setup_queue(chan, config) do
    {xch, xch_type, xch_opts} = normalize_exchange(config[:exchange])
    {q, q_opts} = normalize_queue(config[:queue])

    dl_args = setup_dead_letter(chan, config)
    q_opts = Keyword.update(q_opts, :arguments, dl_args, &(dl_args ++ &1))

    {:ok, %{queue: q}} = AMQP.Queue.declare(chan, q, q_opts)

    # the exchange "" is the default exchange and cannot be declared this way
    unless xch == "" do
      :ok = AMQP.Exchange.declare(chan, xch, xch_type, xch_opts)
      :ok = AMQP.Queue.bind(chan, q, xch, routing_key: config[:routing_key])
    end

    :ok = AMQP.Basic.qos(chan, prefetch_count: config[:prefetch_count])
    {:ok, _} = AMQP.Basic.consume(chan, q, nil, consumer_tag: config[:consumer_tag])
    :ok
  end

  defp setup_dead_letter(chan, config) do
    if config[:dead_letter] do
      {xch, xch_type, xch_opts} = normalize_exchange(config[:dead_letter_exchange])
      {q, q_opts} = normalize_queue(config[:dead_letter_queue])

      {:ok, %{queue: q}} = AMQP.Queue.declare(chan, q, q_opts)

      :ok = AMQP.Exchange.declare(chan, xch, xch_type, xch_opts)
      :ok = AMQP.Queue.bind(chan, q, xch)

      [
        {"x-dead-letter-exchange", :longstr, xch},
        {"x-dead-letter-routing-key", :longstr, config[:dead_letter_routing_key]}
      ]
    else
      []
    end
  end

  defp module_config(module) do
    config = Keyword.merge(@defaults, module.config())
    {q, q_opts} = Keyword.fetch!(config, :queue) |> normalize_queue()
    {xch, xch_type, xch_opts} = Keyword.fetch!(config, :exchange) |> normalize_exchange()
    dl_xch_name = String.replace_prefix("#{xch}.dead-letter", ".", "")
    # by default for the dead-letter exchange use the same type and the same `:durable` option
    # as for the main exchange. When the `:durable` option isn't set explicitly, we determine it
    # based on the main exchange. The default exchange `""` is always durable.
    dl_xch = {dl_xch_name, xch_type, durable: Keyword.get(xch_opts, :durable, xch == "")}
    # by default for the dead-letter queue use the same `:durable` option as for the main queue
    dl_q = {"#{q}_error", durable: Keyword.get(q_opts, :durable, false)}

    config
    |> Keyword.put_new(:routing_key, q)
    |> Keyword.put_new(:dead_letter_routing_key, q)
    |> Keyword.put_new(:dead_letter_exchange, dl_xch)
    |> Keyword.put_new(:dead_letter_queue, dl_q)
    |> Keyword.put_new(:consumer_tag, to_string(module))
  end

  defmacro __using__(opts \\ []) do
    quote location: :keep do
      use GenServer

      import RMQ.Utils, only: [ack: 3, ack: 2, reply: 4, reply: 3]
      import AMQP.Basic, only: [reject: 3, reject: 2, publish: 5, publish: 4]

      @behaviour RMQ.Consumer
      @config unquote(opts)

      def start_link(opts), do: RMQ.Consumer.start_link(__MODULE__, opts)

      @impl RMQ.Consumer
      def setup_queue(chan, config), do: RMQ.Consumer.setup_queue(chan, config)

      @impl RMQ.Consumer
      def config, do: @config

      @impl GenServer
      def init(arg), do: RMQ.Consumer.init(__MODULE__, arg)

      @impl GenServer
      def handle_info(msg, state), do: RMQ.Consumer.handle_info(__MODULE__, msg, state)

      @impl GenServer
      def terminate(reason, state), do: RMQ.Consumer.terminate(__MODULE__, reason, state)

      defoverridable config: 0, setup_queue: 2
    end
  end
end
